cmd/check/main.go

```diff
@@ -49,66 +46,259 @@ func main() {
+
+			pre := ver.Prerelease()
+			if pre != "" {
+				// pre-releases not enabled; skip
+				if !source.PreReleases {
+					continue
+				}
+
+				// contains additional variant
+				if strings.Contains(pre, "-") {
+					continue
+				}
+
+				if !strings.HasPrefix(pre, "alpha.") &&
+					!strings.HasPrefix(pre, "beta.") &&
+					!strings.HasPrefix(pre, "rc.") {
+					// additional variant, not a prerelease segment
+					continue
+				}
+			}
+
+			if cursorVer != nil && (cursorVer.GreaterThan(ver) || cursorVer.Equal(ver)) {
+				// optimization: don't bother fetching digests for lesser (or equal but
+				// less specific, i.e. 6.3 vs 6.3.0) version tags
+				continue
+			}
+		}
+
+		tagRef := repo.Tag(identifier)
+
+		digestImage, err := remote.Image(tagRef, imageOpts...)
+		if err != nil {
+			return resource.CheckResponse{}, fmt.Errorf("get tag digest: %w", err)
+		}
+
+		digest, err := digestImage.Digest()
+		if err != nil {
+			return resource.CheckResponse{}, fmt.Errorf("get cursor image digest: %w", err)
+		}
+
+		tagDigests[identifier] = digest.String()
+
+		if ver != nil {
+			versionTags[ver] = tagRef
+
+			existing, found := digestVersions[digest.String()]
+			if !found || strings.Count(ver.Original(), ".") > strings.Count(existing.Original(), ".") {
+				digestVersions[digest.String()] = ver
+			}
+
+			versions = append(versions, ver)
+		}
+
+		if from != nil && identifier == from.Tag && digest.String() == from.Digest {
+			// if the 'from' version exists and has the same digest, treat its
+			// version as a cursor in the tags, only considering newer versions
+			//
+			// note: the 'from' version will always be the first one hit by this loop
+			cursorVer = ver
+		}
+	}
+
+	sort.Sort(semver.Collection(versions))
+
+	var tagVersions TagVersions
+	for digest, version := range digestVersions {
+		tagVersions = append(tagVersions, TagVersion{
+			TagName: versionTags[version].TagStr(),
+			Digest:  digest,
+			Version: version,
+		})
+	}
+
+	sort.Sort(tagVersions)
+
+	response := resource.CheckResponse{}
+
+	for _, ver := range tagVersions {
+		response = append(response, resource.Version{
+			Tag:    ver.TagName,
+			Digest: ver.Digest,
+		})
+	}
+
+	if latestTag != "" {
+		digest := tagDigests[latestTag]
+
+		_, existsAsSemver := digestVersions[digest]
+		if !existsAsSemver {
+			response = append(response, resource.Version{
+				Tag:    latestTag,
+				Digest: digest,
+			})
+		}
+	}
+
+	return response, nil
+}
+
+type TagVersion struct {
+	TagName string
+	Digest  string
+	Version *semver.Version
+}
+
+type TagVersions []TagVersion
+
+func (vs TagVersions) Len() int           { return len(vs) }
+func (vs TagVersions) Less(i, j int) bool { return vs[i].Version.LessThan(vs[j].Version) }
+func (vs TagVersions) Swap(i, j int)      { vs[i], vs[j] = vs[j], vs[i] }
+
+func checkTag(source resource.Source, version *resource.Version) (resource.CheckResponse, error) {
+	repo, err := name.NewRepository(source.Repository, name.WeakValidation)
```

Comment:
This can be DRY'ed up with checkRepository

Reply:
done!